packages/playwright-test/src/worker/testTracing.ts

```diff
@@ -0,0 +1,171 @@
+            sourceFiles.add(frame.file);
+        }
+      }
+      for (const sourceFile of sourceFiles) {
+        await fs.promises.readFile(sourceFile, 'utf8').then(source => {
+          zipFile.addBuffer(Buffer.from(source), 'resources/src@' + calculateSha1(sourceFile) + '.txt');
+        }).catch(() => {});
+      }
+    }
+
+    const sha1s = new Set<string>();
+    for (const event of this._traceEvents.filter(e => e.type === 'after') as trace.AfterActionTraceEvent[]) {
+      for (const attachment of (event.attachments || [])) {
+        let contentPromise: Promise<Buffer | undefined> | undefined;
+        if (attachment.path)
+          contentPromise = fs.promises.readFile(attachment.path).catch(() => undefined);
+        else if (attachment.base64)
+          contentPromise = Promise.resolve(Buffer.from(attachment.base64, 'base64'));
+
+        const content = await contentPromise;
+        if (content === undefined)
+          continue;
+
+        const sha1 = calculateSha1(content);
+        attachment.sha1 = sha1;
+        delete attachment.path;
+        delete attachment.base64;
+        if (sha1s.has(sha1))
+          continue;
+        sha1s.add(sha1);
+        zipFile.addBuffer(content, 'resources/' + sha1);
+      }
+    }
+
+    const traceContent = Buffer.from(this._traceEvents.map(e => JSON.stringify(e)).join('\n'));
+    zipFile.addBuffer(traceContent, 'trace.trace');
+
+    await new Promise(f => {
+      zipFile.end(undefined, () => {
+        zipFile.outputStream.pipe(fs.createWriteStream(fileName)).on('close', f);
+      });
+    });
+  }
+
+  appendStdioToTrace(type: 'stdout' | 'stderr', chunk: string | Buffer) {
+    this._appendTraceEvent({
+      type,
+      timestamp: monotonicTime(),
+      text: typeof chunk === 'string' ? chunk : undefined,
+      base64: typeof chunk === 'string' ? undefined : chunk.toString('base64'),
+    });
+  }
+
+  appendBeforeActionForStep(callId: string, parentId: string | undefined, apiName: string, params: Record<string, any> | undefined, wallTime: number, stack: StackFrame[]) {
+    this._appendTraceEvent({
+      type: 'before',
+      callId,
+      parentId,
+      wallTime,
+      startTime: monotonicTime(),
+      class: 'Test',
+      method: 'step',
+      apiName,
+      params: Object.fromEntries(Object.entries(params || {}).map(([name, value]) => [name, generatePreview(value)])),
+      stack,
+    });
+  }
+
+  appendAfterActionForStep(callId: string, attachments: Attachment[], initialAttachments: Set<Attachment>, error?: SerializedError['error']) {
+    this._appendTraceEvent({
+      type: 'after',
+      callId,
+      endTime: monotonicTime(),
+      log: [],
+      attachments: serializeAttachments(attachments, initialAttachments),
+      error,
+    });
+  }
+
+  private _appendTraceEvent(event: trace.TraceEvent) {
+    this._traceEvents.push(event);
+    if (this._liveTraceFile)
+      fs.appendFileSync(this._liveTraceFile, JSON.stringify(event) + '\n');
```

Comment:
Isn't this called on every playwright api call, because each call is a step?

Reply:
It is!